### ISSUE TITLE
Route to custom libraries under node modules

### ISSUE THREAD

**Issue by abrahamq** [urbanriskmap/petabencana.id#12]
Can't access css assets for leaflet in node_modules/ from html. Can do it from from javascript router, but not from html. Figure out how to do this nicely. Right now we're copying the leaflet folder into the aurelia src/ folder.

**Comment by mgvarley** [comment]
Hi @abrahamq @MaanasaPriyaa - just worked out how to do this:

1. `npm install leaflet` in root (looks like you have already done this)
2. Configure aurelia with leaflet as a dependency.  This will bundle leaflet into the vendor-bundle.js file that aurelia creates and also make it available to use in development.  Since leaflet now bundles the css as part of the dist this needs to be explicitly referenced.  (aurelia team are working on a more elegant cli to do this in future).  In the meantime this will do it:

`
...
{
  "name": "leaflet",
  "path": "../node_modules/leaflet/dist",
  "main": "leaflet",
  "resources": [
    "leaflet.css"
  ]
},
...
`

3. In your map.html (or where you are showing your map), include a require tag and then initialise your map div, e.g.:

`
<template>
  <require from="leaflet/leaflet.css"></require>
  <div class="map text-center">
    <div id="mapid" style="height: 800px, width:100%"></div>
  </div>
</template>
`

I have included inline style here for simplicity but you will probably want a specific map.css file in the map/ folder to keep things clean.  The key thing with any leaflet is to set the height to a specific value, 100% doesn't work.  If you want a full screen map that auto updates when the window resizes you will need to include some js to handle this using window DOM object - I can help if you need as am doing this in React, should be similar in aurelia

4. In your map.js file create an attached() method, this tells aurelia to do some stuff once the view is ready (a bit like $.ready()) e.g.

`
  attached() {
    let map = L.map('mapid').setView([51.505, -0.09], 13);
    let urlTemplate = 'http://{s}.tile.osm.org/{z}/{x}/{y}.png';
    map.addLayer(L.tileLayer(urlTemplate, { minZoom: 4 }));
  }
`

Hope this helps!

**Comment by mgvarley** [comment]
The same approach can be taken for any client side dependency that you need, it is simpler for dependencies that don't have special layout or bundle css.  There is a good guide that explains more here.  Makes sense once you get your head around it (it took me a while!) http://aurelia.io/hub.html#/doc/article/aurelia/framework/latest/the-aurelia-cli/6

**Comment by abrahamq** [comment]
awesome fix! Thanks for the super quick resolution. 
Here's a broader question: Do we need to think about a front end package manager other than npm? Why do people use bower/browserify? Would a tool like those make this simpler? 

Thanks!

**Comment by mgvarley** [comment]
No problem Abe glad we could get this up and running.  You should not need bower / browserify, these are useful in the context where you are building a more traditional web application comprised of a series of html pages and want to manage the third party dependencies which you would previously have to download and store in an assets (or equivalent) folder or reference to their respective CDN.  With single page frameworks such aurelia there is just a single html file which then hands off to compiled javascript which contains the actual application, including dependencies.  Aurelia manages this compilation and build process for you, you just need to give it some help in deciding which libraries to bring in.


The official guide in the link I included references npm although I have also seen references to jspm, (http://jspm.io/
) a newer alternative which I believe also allows installing dependencies directly from git repos.  I did not introduce that in the PoC as it would be more one more thing to learn.  The npm approach seems reasonable although a bit of a faff in terms of the aurelia.config part, they do say they are going to address this in a subsequent version of the CLI.  


Mark




>

**Comment by abrahamq** [comment]
Awesome, thanks! That makes sense.
